pkg/controllers/upgrade_reconcile.go

```diff
@@ -0,0 +1,271 @@
+	if err := u.PFacts.Collect(ctx, u.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if ok, err := u.isUpgradeNeeded(ctx); !ok || err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Functions to perform upgrade processing.  Order matters.
+	funcs := []func(context.Context) (ctrl.Result, error){
+		// Initiate an upgrade by setting condition and event recording
+		u.startUpgrade,
+		// Do a clean shutdown of the cluster
+		u.stopCluster,
+		// Delete the sts objects.  We don't rely on k8s rolling upgrade.
+		// Everything must be destroyed then regenerated.
+		u.deleteStatefulSets,
+		// Create the sts object with the new image name.
+		u.recreateStatefulSets,
+		// Start up vertica in each pod.
+		u.restartCluster,
+		// Cleanup up the condition and event recording for a completed upgrade
+		u.finishUpgrade,
+	}
+	for _, fn := range funcs {
+		if res, err := fn(ctx); res.Requeue || err != nil {
+			return res, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// startUpgrade handles condition status and event recording for start of upgrade
+func (u *UpgradeReconciler) startUpgrade(ctx context.Context) (ctrl.Result, error) {
+	u.Log.Info("Starting upgrade for reconciliation iteration", "ContinuingUpgrade", u.ContinuingUpgrade)
+	if err := u.toggleUpgradeInProgress(ctx, corev1.ConditionTrue); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// We only log an event message the first time we begin an upgrade.
+	if !u.ContinuingUpgrade {
+		u.VRec.EVRec.Eventf(u.Vdb, corev1.EventTypeNormal, events.UpgradeStart,
+			"Vertica server upgrade has started.  Upgrading to image '%s'", u.Vdb.Spec.Image)
+	}
+	return ctrl.Result{}, nil
+}
+
+// finishUpgrade handles condition status and event recording for the end of an upgrade
+func (u *UpgradeReconciler) finishUpgrade(ctx context.Context) (ctrl.Result, error) {
+	if err := u.toggleUpgradeInProgress(ctx, corev1.ConditionFalse); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	u.VRec.EVRec.Eventf(u.Vdb, corev1.EventTypeNormal, events.UpgradeSucceeded,
+		"Vertica server upgrade has completed successfully")
+
+	return ctrl.Result{}, nil
+}
+
+// isUpgradeNeeded returns true if we are in the middle of an upgrade or we need to start one
+func (u *UpgradeReconciler) isUpgradeNeeded(ctx context.Context) (bool, error) {
+	// We first check if the status condition indicates the upgrade is in progress
+	inx, ok := vapi.VerticaDBConditionIndexMap[vapi.UpgradeInProgress]
+	if !ok {
+		return false, fmt.Errorf("verticaDB condition '%s' missing from VerticaDBConditionType", vapi.UpgradeInProgress)
+	}
+	if inx < len(u.Vdb.Status.Conditions) && u.Vdb.Status.Conditions[inx].Status == corev1.ConditionTrue {
+		// Set a flag to indicate that we are continuing an upgrade.  This silences the UpgradeStarted event.
+		u.ContinuingUpgrade = true
+		return true, nil
+	}
+
+	// Next check if an upgrade is needed based on the image being different
+	// between the Vdb and any of the statefulset's.
```

Comment:
The image being different meaning image name/tag? If so, I'm wondering if we do anything to prevent "downgrade".

Reply:
Yes, that is correct.  We don't make any distinction between upgrade or downgrade.  If the image name changes, then we follow the upgrade process (shutdown cluster, update image, restart cluster).  We need this in case you ever want to rollback an upgrade.  I would consider renaming it if you think it is confusing?  upgrade => imageChange

Reply:
I see, yes "imageChange" is much more clear to me.

Reply:
Maybe I am nitpicking here but is there anything we could do to differentiate rollback-based downgrade and direct downgrade? Vertica doesn't support downgrade and that might lead to things breaking.

Reply:
Unfortunately, there is nothing we can do for this in the operator.  There isn't any notion of binary version in the catalog.  This likely needs engine support to have some mechanism to know when we have downgraded.

Reply:
I see. Thanks!